### ISSUE TITLE
Email Validator does not accept + and longed domain names while it is allowed in ICANN

### ISSUE THREAD

**Issue by racer1988** [ESAPI/esapi-java-legacy#374]
while doing some email validation on our Java Backend with @0xnm

we notice that:

```
Validator.Email=^[A-Za-z0-9._%'-]+@[A-Za-z0-9.-]+\\.[a-zA-Z]{2,4}$
```

has two issues and does not respect these rules:
https://www.owasp.org/index.php/Input_Validation_Cheat_Sheet#Email_Validation_Basics

1) Domain name is limited from 2 to 4 characters
ICANN now removed limitation for domain names, so correct regex should be {2,}

2) This does not allow + sign inside the email
however

```
info+user@something.com
```

is a valid alias email (gmail uses this for aliases for example)

There is a reason not to include plus sign in the email validator?
Is this going to be changed in a new version?

**Comment by xeno6696** [comment]
You should keep sane upper boundaries.  Yes, ICANN specifies a non-determinant length, but we wouldn't encourage your suggestion of `{2,}`.  (Well, I wouldn't... but I doubt any other project contributors would disagree.)  

We could update the regex however to account for the gmail use-case, but I think it would be a regex specific to gmail because I don't think that syntax is in any RFC that I'm aware of.  

@kwwall, make this an enhancement and add a "gmail_email" regex in validation.properties for the next point release?

**Comment by racer1988** [comment]
Yes I understand the sane upper boundaries idea.

Unlimited is probably too much, but a high limit following ICANN would be good.

If I am not mistaken DNS allows for a maximum of 63 characters for an individual label, so probably 63 is the more accurate upper boundary 

What do you think?

**Comment by xeno6696** [comment]
According to [RFC 1035](https://tools.ietf.org/html/rfc1035)  section 2.3.4 the length is 255 octets.  But since Microsoft introduces oddball complexities... and I *have* seen prod applications deployed to IIS servers, Microsoft would have to be a consideration.  From [this Microsoft ](https://blogs.msdn.microsoft.com/oldnewthing/20120412-00/?p=7873)blog:


> If you sit down and do the math, you’ll see that the the readable maximum length of an ASCII DNS name is 253 characters: You don’t encode the dots, but you do encode the length bytes, so they cancel out, except for the length byte of the first label and the length byte of the root label, for an additional cost of two bytes. (On the off chance that you explicitly specified the root label, don’t count it towards the 253-character limit.)
> 
> If you use UTF-8 encoding, then the maximum length is harder to describe since UTF-8 is a variable-length encoding.

Given this, I think to be safe we specify 62.  @kwwall , any input here?

**Comment by xeno6696** [comment]
Apparently I fixed this in commit 32f5e84dd8c06e

My apologies,  We created a new "Gmail" regex that takes the `+` character, and we also limited the domain size to 62 per discussion here.
